### ISSUE TITLE
Scheduler - Virtual Scrolling

### ISSUE THREAD

**Issue by AntonSermyazhko** [DevExpress/DevExtreme#14285]
# Scheduler - Virtual Scrolling

# The Problem

Scheduler users may experience performance issues if a Scheduler view contains many resources, groups, or appointments.

# The Proposed Solution

We will implement Virtual Scrolling. In this mode, our Scheduler renders only visible UI elements within a view. Once a UI element leaves the visible view area, the Scheduler removes it from the DOM. This reduces the DOM size and significantly improves performance.
 
The Virtual Scrolling API allows developers to do the following:
 
- Set the scrolling mode for the whole widget:
 
```js
    scrolling: { mode:  'standard' || 'virtual' } 
```
 
- Set the scrolling mode for an individual view:
  
```js
    views: [ { …, scrolling: { mode:  'standard' || 'virtual' } } ]
```
 
`mode`'s value can be either of the following strings:
 
- `standard` - rows and appointments are rendered simultaneously.
 
- `virtual` - rows and appointments are rendered when they get into the viewport and removed once they leave it.

# Try It

## Live Sandboxes

[Virtual Scrolling sample](https://codepen.io/pin361/pen/KKzoyzo)

# We Need Your Feedback

## Take a Quick Poll

[Do you find virtual scrolling in the Scheduler useful](https://forms.gle/oaEJ1syMpVLvFhTr9)?

# Get Notified of Updates

Subscribe to this thread or to our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by AntonSermyazhko** [comment]
We have updated the discussion page with a [live sandbox sample](https://codepen.io/pin361/pen/KKzoyzo).

**Comment by radomeit** [comment]
Hi, basically this feature is great and will allow us to deliver a much smoother scheduler experience to our users.
Since we only use the timeline view mode, I modified and tested your example with it, but it seems that virtual scrolling doesn't  work here (it takes some time before the scheduler shows up, probably cause its rendering all appointments) and also the vertical scrollbars are hidden / you can't scroll down, so you can see only 8 resources instead of 30: https://codepen.io/radomeit/pen/vYKBjBM
Did I miss sth.?

**Comment by AntonSermyazhko** [comment]
> Hi, basically this feature is great and will allow us to deliver a much smoother scheduler experience to our users.
> Since we only use the timeline view mode, I modified and tested your example with it, but it seems that virtual scrolling doesn't work here (it takes some time before the scheduler shows up, probably cause its rendering all appointments) and also the vertical scrollbars are hidden / you can't scroll down, so you can see only 8 resources instead of 30: [https://codepen.io/radomeit/pen/vYKBjBM](https://codepen.io/radomeit/pen/vYKBjBM)
> Did I miss sth.?

Hi,
Currently, the Scheduler widget supports the Virtual Scrolling functionality only in the 'day', 'week', 'workWeek' views. We will reflect this limitation in our documentation. We have plans to support this functionality for timeline views. However, it is unlikely we add it in v20.2.

**Comment by bbarber9** [comment]
We were forced to use another scheduler component due to the poor performance of the timeline view. This would be game changing in the timeline view for us.

**Comment by MaximKudriavtsev** [comment]
Thanks to everybody who gave us early feedback on this feature. The feature is now available in the [v20.2 release](https://github.com/DevExpress/DevExtreme/releases/tag/20.2.3).

Currently, the Scheduler component supports the [Virtual Scrolling functionality](https://js.devexpress.com/Demos/WidgetsGallery/Demo/Scheduler/VirtualScrolling/jQuery/Light/) only in the `day`, `week`, `workWeek` views, and with vertical grouping. We will add this feature to Timelines, and it will work with [horizontal grouping](https://supportcenter.devexpress.com/ticket/details/t936668/scheduler-virtual-scrolling-does-not-work-when-horizontal-grouping-is-enabled) in future releases.

I'm closing this thread. In the case of bugs or questions, feel free to create new [GitHub issues](https://github.com/DevExpress/DevExtreme/issues) or tickets in our [Support Center](https://www.devexpress.com/Support/Center/).

**Comment by radomeit** [comment]
Hi @MaximKudriavtsev , can you give any time schedule, when this great functionality will be available for the timeline view as well?

**Comment by MaximKudriavtsev** [comment]
Hi @radomeit,

This feature is a high priority for us. We are working on it right now. Virtual scrolling in timelines will be available in v21.1 (June-July 2021). I will update this topic once we have any news.

**Comment by MrTrulsen** [comment]
Hi, 
Will virtual scrolling ever be supported in the monthly view of the Scheduler?

**Comment by MaximKudriavtsev** [comment]
Hi @MrTrulsen,

Yes, it will. We plan to release the Virtual Scrolling feature for all Scheduler views in v21.1 (June-July 2021). I will update this topic once this feature is available in early access.

**Comment by MaximKudriavtsev** [comment]
Hi everybody,

We are happy to announce that Virtual Scrolling in horizontal views is available in DevExtreme NPM distribution (in 21.1 [Sprint 9](https://github.com/DevExpress/DevExtreme/releases/tag/21.1.1-sprint.9) and newer).

You can use it in all available Scheduler views. To enable the feature, set the `scrolling` property to `{ mode: 'virtual' }`.

Please try it and let us know if any issues appear. As always, we'd love to hear your feedback.



### NPM Packages
```
devextreme@21.1.1-alpha-21075-0317

devextreme-angular@21.1.1-alpha-21075-0317
devextreme-react@21.1.1-alpha-21075-0317
devextreme-vue@21.1.1-alpha-21075-0317
```

**Comment by ksercs** [comment]
Thank you everybody, who gave us feedback on this feature. The feature is available in v21.1. I'm closing this thread. In the case of bugs or questions, feel free to create a new ticket in our [Support Center](https://www.devexpress.com/Support/Center/).
